python/graphframes/examples/belief_propagation.py

```diff
@@ -0,0 +1,173 @@
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import math
+
+from pyspark import SparkConf, SparkContext
+from pyspark.sql import SQLContext, functions as sqlfunctions, types
+
+from graphframes import GraphFrame, AggregateMessages as AM
+
+__all__ = ['BeliefPropagation']
+
+
+class BeliefPropagation(object):
+    """Example code for Belief Propagation (BP)
+
+    This provides a template for building customized BP algorithms for different types of graphical
+    models.
+
+    This example:
+
+    * Ising model on a grid
+    * Parallel Belief Propagation using colored fields
+
+    Ising models are probabilistic graphical models over binary variables
+    (see :meth:`Graphs.gridIsingModel()`).
+
+    Belief Propagation (BP) provides marginal probabilities of the values of the variables
+    x\ :sub:`i` i.e., P(x\ :sub:`i`) for each i.  This allows a user to understand likely values of
+    variables. See `Wikipedia <https://en.wikipedia.org/wiki/Belief_propagation>`__ for more
+    information on BP.
+
+    We use a batch synchronous BP algorithm, where batches of vertices are updated synchronously.
+    We follow the mean field update algorithm in Slide 13 of the
+    `talk slides <http://www.eecs.berkeley.edu/~wainwrig/Talks/A_GraphModel_Tutorial>`__ from:
+    Wainwright. "Graphical models, message-passing algorithms, and convex optimization."
+
+    The batches are chosen according to a coloring. For background on graph colorings for
+    inference, see for example: Gonzalez et al. "Parallel Gibbs Sampling: From Colored Fields to
+    Thin Junction Trees." AISTATS, 2011.
+
+    The BP algorithm works by:
+
+    * Coloring the graph by assigning a color to each vertex such that no neighboring vertices
+      share the same color.
+    * In each step of BP, update all vertices of a single color.  Alternate colors.
+     """
+
+    @classmethod
+    def runBPwithGraphFrames(cls, g, numIter):
+        """Run Belief Propagation using GraphFrame.
+
+        This implementation of BP shows how to use GraphFrame's aggregateMessages method.
+        """
+        # choose colors for vertices for BP scheduling
+        colorG = cls._colorGraph(g)
+        numColors = colorG.vertices.select('color').distinct().count()
+
+        # TODO: handle vertices without any edges
```

Comment:
This comment is also in the scala side.
In our example, using the gridIsingModel example graph, the graph should not contain such vertices anyway.